Make kubelet set alpha.kubernetes.io/provided-node-ip unconditionally

mdbooth opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

<!--
Add one of the following kinds:
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
The easier of two possible fixes for an issue where node addresses flap during an upgrade to an external CCM. This fix causes kubelet to apply the alpha.kubernetes.io/provided-node-ip annotation unconditionally (not only when --cloud-provider=external). This does not exclude a future fix involving the root cause of the issue, which is that kubelet and cloud-controller-manager both attempt to manage node addresses when --cloud-provider is not set to external.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #109793

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The node annotation alpha.kubernetes.io/provided-node-ip is no longer set ONLY when `--cloud-provider=external`.  Now, it is set on kubelet startup if the `--cloud-provider` flag is set at all, including the deprecated in-tree providers.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

mdbooth committed:
Make kubelet set alpha.kubernetes.io/provided-node-ip unconditionally

elmiko commented on pkg/kubelet/nodestatus/setters.go:
```diff
@@ -94,14 +94,21 @@ func NodeAddress(nodeIPs []net.IP, // typically Kubelet.nodeIPs
+		// node controller in the cluster but kubelet is still running
+		// the in-tree provider. Adding this annotation in all cases
+		// ensures that while Addresses flap between the competing
+		// controllers, they at least flap consistently.
```
nice comment, thanks Matt

elmiko reviewed (commented):
i am not super familiar with code here, but this generally makes sense to me
/lgtm

nckturner commented:
/triage accepted

nckturner commented:
/lgtm

nckturner commented:
cc @kubernetes/sig-node-pr-reviews

nckturner commented:
/assign @yujuhong

nckturner commented:
ping @yujuhong @kubernetes/sig-node-bugs @kubernetes/sig-node-pr-reviews, who is the correct reviewer/approver from sig-node for this?

dchen1107 commented:
Is this a regression from 1.23? Do we need to cherrypick?

nckturner commented:
@dchen1107 It's not a regression, but it is a bug.  I think we should definitely cherry pick.

aojea commented:
/assign @danwinship @andrewsykim 

hmm, changing the behavior to ALWAYS annotate just to solve one corner and specific case seems excessive to me 
Why clusters without any cloud provider at all has to carry with a cloud provider annotation, that is also alpha by the way, and this will make it harder to move in the future

aojea commented:
/sig network

danwinship commented:
> Why clusters without any cloud provider at all has to carry with a cloud provider annotation

The patch would work if it only annotated in clusters where `cloudProvider != ""` (ie, external or legacy cloud provider, but not bare-metal).

> that is also alpha by the way, and this will make it harder to move in the future

It's more like a GA annotation with "alpha" in its name :slightly_frowning_face:. Though yeah, using it more would make it harder to fix that.

aojea commented:
> The patch would work if it only annotated in clusters where `cloudProvider != ""` (ie, external or legacy cloud provider, but not bare-metal).

ah, ok, excuse me then

danwinship commented:
I didn't check the code. The fix may not literally be `cloudProvider != ""` but that _idea_ should work

mdbooth commented:
I have updated the PR with the request from @aojea to make no change for the non cloud-provider case.

I have also added unit tests which verify this behaviour.

mdbooth commented:
Failures both look like flakes

/retest-required

nckturner commented:
Tests passed @mdbooth @aojea

elmiko reviewed (commented):
i think this is looking good, thanks for the updates @mdbooth 
/lgtm

aojea commented:
> > ```
> > if externalCloudProvider || cloud != nil {
> >      do annotation
> > } 
> > ```
> 
> I'll make this change.

thanks for your comprehension, defer approval for people more familiar with cloud-provider codebase

/assign @cheftako @thockin @andrewsykim

markyjackson-taulia commented:
Hi @mdbooth  My name is Marky Jackson and I am one of the k8s 1.25 bug triage shadow assigned to track this body of work. Just checking in to see if this is still on track for k8s 1.25?

mdbooth commented:
> Hi @mdbooth  My name is Marky Jackson and I am one of the k8s 1.25 bug triage shadow assigned to track this body of work. Just checking in to see if this is still on track for k8s 1.25?

I certainly hope so! Nothing to do from my side that I'm aware of beyond soliciting approvals.

dchen1107 commented:
Another PR I came across might try to address a similar issue caused by the external cloud provider migration: https://github.com/kubernetes/kubernetes/pull/110704 cc @haiqianz

dchen1107 commented on pkg/kubelet/nodestatus/setters.go:
```diff
@@ -94,14 +94,28 @@ func NodeAddress(nodeIPs []net.IP, // typically Kubelet.nodeIPs
 
-		if externalCloudProvider {
+		if externalCloudProvider || cloud != nil {
+			// Annotate the Node object with nodeIP for external cloud provider.
```
You are changing the semantic of the pre-defined annotation: https://kubernetes.io/docs/reference/labels-annotations-taints/#alpha-kubernetes-io-provided-node-ip. Change the related doc?

dchen1107 reviewed (commented):
One nit on the related doc. Otherwise LGTM

dchen1107 commented:
/lgtm

nckturner commented:
/lgtm

Who can approve?  @dchen1107?

aojea commented:
I just found there is a KEP to handle this in/out tree cloud provider migration problem https://github.com/kubernetes/enhancements/tree/master/keps/sig-cloud-provider/2436-controller-manager-leader-migration , can not be used for this?

mdbooth commented:
> I just found there is a KEP to handle this in/out tree cloud provider migration problem https://github.com/kubernetes/enhancements/tree/master/keps/sig-cloud-provider/2436-controller-manager-leader-migration , can not be used for this?

Unfortunately not. We're effectively migrating the 'Node controller' from running locally in every kubelet to running centrally in a single Node controller. We'd need a separate leader election for each individual kubelet.

If we wanted to add more annotations I suppose the kubelet could annotate the Node to indicate who is expected to manage it, and the Node controller could be taught to ignore kubelet-managed Nodes. However, this annotation would have an extremely limited useful lifetime.

nckturner commented:
This PR involves the kubelet and the CCM, (the above KEP refers to a lock between the KCM and CCM) and is specific to the period of time when migration is taking place (old kubelets are still running and reconciling Node Addresses, and the new CCM also begins reconciling node addresses).  In order for the above leader migration lock to be used for this, kubelet would need to take that migration lock before doing its node address management. Might be possible, but I think regardless we should still make the above change, because they are not mutually exclusive and it is a simple remediation to the flapping address behavior that is described above.

dchen1107 commented:
@nckturner thanks for open pr for doc updates: https://github.com/kubernetes/website/pull/35027

/lgtm
/approve

aojea commented:
> @nckturner I recommend adding a changelog entry (not `NONE`) for this change. The updated behavior is visible to end users.

agreed

aojea commented:
/release-note-edit ```release-note The node annotation alpha.kubernetes.io/provided-node-ip is no longer set ONLY when `--cloud-provider=external`.  Now, it is set on kubelet startup if the `--cloud-provider` flag is set at all, including the deprecated in-tree providers. ```
